databroker/broker.py

```diff
@@ -562,3 +566,130 @@ def _munge_time(t, timezone):
     """
     t = datetime.fromtimestamp(t)
     return timezone.localize(t).replace(microsecond=0).isoformat()
+
+
+known_special_keys = {
+    # The duration of the scan
+    'duration': lambda header: (datetime.fromtimestamp(header.stop.time) -
+                                datetime.fromtimestamp(header.start.time)),
+    # The number of events per event descriptor
+    'num_events': lambda header: [
+        len(list(mdsc.get_events_generator(descriptor, False)))
+        for descriptor in header.descriptors],
+}
+
+def _get_uid(key, document):
+    try:
+        # see if we have a hyphen
+        key, num = key.split('-')
+        num = int(num)
+    except ValueError:
+        # Nope! Must be asking for the whole uid
+        num = None
+    return document.uid[:num]
+
+
+@singledispatch
+def _get_value(h, key, default_value):
+    """
+    Get the value specified by `key` from the Header `h`. Give the value
+    of `default_value` for keys that are not in `h`
+
+    Parameters
+    ----------
+    h : databroker.core.Header
+        A single header to try and get a key from
+    key : str
+        The key to find a value for in `h`. See the `*keys` section of the
+        `databroker.broker.summarize` docstring
+    default_value : str
+        The default value for which to fill missing keys
+
+    Returns
+    -------
+    object
+        `default_value` or whatever `key` would return.
+    """
+    raise NotImplementedError("_get_value is not implemented for type(h)={}"
+                              "".format(h))
+
+
+@_get_value.register(Header)
+def _(header, key, default_value):
+    split = key.split('-', maxsplit=1)
+    if len(split) == 1:
+        # `key` is in `known_special_keys` or is assumed to be a key in the
+        # RunStart document
+        key, = split
+        if key in known_special_keys:
+            return known_special_keys[key](header)
+        # Use the registered handling for `Documents`
+        return _get_value(header.start, key, default_value)
+    s0, s1 = split
+    if 'uid' in s0:
+        return _get_uid(key, header.start)
+    if s0 == 'start':
+        return _get_value(header.start, s1, default_value)
+    elif s0 == 'stop':
+        return _get_value(header.stop, s1, default_value)
+    elif s0 == 'descriptor' or s0 == 'descriptors':
+        return [_get_value(descriptor, s1, default_value)
+                for descriptor in header.descriptors]
+
+
+@_get_value.register(Document)
+def _(header, key, default_value):
+    document = header
+    if 'uid' in key:
+        return _get_uid(key, document)
+    if key == 'time':
+        return datetime.fromtimestamp(document.get(key, default_value))
+    return document.get(key, default_value)
+
+
+def summarize(headers, *keys, default_value="N/A"):
```

Comment:
This is not valid LPy.  Need to use kwrags, pop the one key you want and then check that it is empty.